File: examples/kotlin/examples/ClientCollectionUtils.kt
package io.ekodb.client.examples

import io.ekodb.client.EkoDBClient
import io.ekodb.client.types.Record
import io.github.cdimascio.dotenv.dotenv
import kotlinx.coroutines.runBlocking

/**
 * Collection Utilities example - Demonstrates collection utility methods:
 * - collectionExists: Check if a collection exists
 * - count: Count documents in a collection
 * - listCollections: List all collections
 * - getCollection: Get collection metadata
 * - deleteCollection: Delete a collection
 */
fun main(): Unit = runBlocking {
    val dotenv = dotenv()
    val baseUrl = dotenv["API_BASE_URL"] ?: "http://localhost:8080"
    val apiKey = dotenv["API_BASE_KEY"] ?: "a-test-api-key-from-ekodb"

    val client = EkoDBClient.builder()
        .baseUrl(baseUrl)
        .apiKey(apiKey)
        .build()

    val collection = "collection_utils_test_kt"
    val nonexistentCollection = "nonexistent_collection_xyz_${System.nanoTime()}"
    var failure: Throwable? = null

    println("=== ekoDB Kotlin Client - Collection Utilities Example ===\n")

    try {
        // Make the example deterministic across repeat runs while keeping the
        // collectionExists calls below as the behavior being demonstrated.
        if (client.collectionExists(collection)) {
            client.deleteCollection(collection)
        }

        // Example 1: Check if collection exists (should be false initially)
        println("=== Check Collection Exists (Before Creation) ===")
        try {
            val exists = client.collectionExists(collection)
            println("Collection '$collection' exists: $exists")
            check(!exists) { "Expected '$collection' not to exist before creation" }
        } catch (e: Exception) {
            println("CollectionExists error: ${e.message}")
            throw e
        }

        // Example 2: Create some test documents to implicitly create the collection
        println("\n=== Creating Test Documents ===")
        for (i in 1..5) {
            val record = Record.new()
                .insert("name", "Test Document $i")
                .insert("index", i)
            try {
                client.insert(collection, record)
            } catch (e: Exception) {
                println("Insert error: ${e.message}")
                throw e
            }
        }
        println("Created 5 test documents")

        // Example 3: Check if collection exists now (should be true)
        println("\n=== Check Collection Exists (After Creation) ===")
        try {
            val exists = client.collectionExists(collection)
            println("Collection '$collection' exists: $exists")
            check(exists) { "Expected '$collection' to exist after creation" }
        } catch (e: Exception) {
            println("CollectionExists error: ${e.message}")
            throw e
        }

        // Example 4: Count documents in collection
        println("\n=== Count Documents ===")
        try {
            val count = client.count(collection)
            check(count == 5L) { "Expected 5 documents in '$collection', found $count" }
            println("Document count in '$collection': $count")
        } catch (e: Exception) {
            println("Count error: ${e.message}")
            throw e
        }

        // Example 5: Get collection metadata
        println("\n=== Get Collection Metadata ===")
        try {
            val metadata = client.getCollection(collection)
            println("Collection metadata: $metadata")
        } catch (e: Exception) {
            println("GetCollection error: ${e.message}")
            throw e
        }

        // Example 6: List all collections
        println("\n=== List Collections ===")
        try {
            val collections = client.listCollections()
            println("All collections (${collections.size}):")
            collections.forEach { println("  - $it") }
        } catch (e: Exception) {
            println("ListCollections error: ${e.message}")
            throw e
        }

        // Example 7: Check non-existent collection
        println("\n=== Check Non-Existent Collection ===")
        try {
            val exists = client.collectionExists(nonexistentCollection)
            println("Collection '$nonexistentCollection' exists: $exists")
            check(!exists) { "Expected '$nonexistentCollection' not to exist" }
        } catch (e: Exception) {
            println("CollectionExists error: ${e.message}")
            throw e
        }
    } catch (error: Throwable) {
        failure = error
    } finally {
        // Cleanup: Delete the test collection
        println("\n=== Cleanup ===")
        try {
            client.deleteCollection(collection)
            println("Deleted collection '$collection'")
        } catch (e: Exception) {
            println("Could not delete collection: ${e.message}")
            val primary = failure
            if (primary == null) failure = e else primary.addSuppressed(e)
        }

        try {
            client.close()
        } catch (closeError: Throwable) {
            val primary = failure
            if (primary == null) failure = closeError else primary.addSuppressed(closeError)
        }
    }
    failure?.let { throw it }
    println("\n=== Collection Utilities Example Complete ===")
}

File: examples/rust/examples/client_collection_utils.rs
//! Collection Utilities Example - Using ekodb_client library
//!
//! This example demonstrates collection utility methods:
//! - collection_exists: Check if a collection exists
//! - count_documents: Count documents in a collection
//! - list_collections: List all collections
//! - delete_collection: Delete a collection

use ekodb_client::{Client, Record};
use std::env;

#[tokio::main]
async fn main() -> Result<(), Box<dyn std::error::Error>> {
    dotenv::dotenv().ok();

    let base_url = env::var("API_BASE_URL").unwrap_or_else(|_| "http://localhost:8080".to_string());
    let api_key =
        env::var("API_BASE_KEY").unwrap_or_else(|_| "a-test-api-key-from-ekodb".to_string());

    // Create ekoDB client
    let client = Client::builder()
        .base_url(&base_url)
        .api_key(&api_key)
        .build()?;

    println!("✓ Client created");

    let collection = "collection_utils_test_rust";

    // Example 1: Check if collection exists (should be false initially)
    println!("\n=== Check Collection Exists (Before Creation) ===");
    let exists = client.collection_exists(collection).await?;
    println!("Collection '{}' exists: {}", collection, exists);

    // Example 2: Create some test documents
    println!("\n=== Creating Test Documents ===");
    for i in 1..=5 {
        let mut record = Record::new();
        record.insert("name", format!("Test Document {}", i));
        record.insert("index", i);
        client.insert(collection, record, None).await?;
    }
    println!("Created 5 test documents");

    // Example 3: Check if collection exists now (should be true)
    println!("\n=== Check Collection Exists (After Creation) ===");
    let exists = client.collection_exists(collection).await?;
    println!("Collection '{}' exists: {}", collection, exists);

    // Example 4: Count documents in collection
    println!("\n=== Count Documents ===");
    let count = client.count_documents(collection).await?;
    println!("Document count in '{}': {}", collection, count);

    // Example 5: List all collections (our test collection should appear)
    println!("\n=== List Collections ===");
    let collections = client.list_collections().await?;
    println!("Total collections: {}", collections.len());
    for name in &collections {
        let marker = if name == collection {
            " <-- our test"
        } else {
            ""
        };
        println!("  - {}{}", name, marker);
    }

    // Example 6: Check non-existent collection
    println!("\n=== Check Non-Existent Collection ===");
    let exists = client
        .collection_exists("nonexistent_collection_xyz")
        .await?;
    println!("Collection 'nonexistent_collection_xyz' exists: {}", exists);

    // Cleanup: Delete the test collection
    println!("\n=== Cleanup ===");
    client.delete_collection(collection).await?;
    println!("Deleted collection '{}'", collection);

    println!("\n✓ Collection Utilities example complete");

    Ok(())
}
